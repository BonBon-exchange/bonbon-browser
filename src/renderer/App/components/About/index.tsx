import { About } from './About';

export * from './About';
export default About;
