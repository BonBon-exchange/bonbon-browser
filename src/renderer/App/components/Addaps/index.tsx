import { Addaps } from './Addaps';

export * from './Addaps';
export default Addaps;
