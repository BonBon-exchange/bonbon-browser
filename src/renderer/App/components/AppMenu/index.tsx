import { AppMenu } from './AppMenu';

export * from './AppMenu';
export default AppMenu;
