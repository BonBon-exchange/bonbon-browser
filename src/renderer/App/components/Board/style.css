.Board__container {
  display: flex;
  height: 300vh;
  width: calc(100% - 50px);
  top: 0px;
  left: 50px;
  position: absolute;
  background: linear-gradient(
    200.96deg,
    #fedc2a -29.09%,
    #dd5789 51.77%,
    #7a2c9e 129.35%
  );
}
