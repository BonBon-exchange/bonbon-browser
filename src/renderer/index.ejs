<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8" />
    <meta
      http-equiv="Content-Security-Policy"
    />
    <title>BonBon</title>
  </head>
  <body>
    <div id="root"></div>
  </body>
</html>
