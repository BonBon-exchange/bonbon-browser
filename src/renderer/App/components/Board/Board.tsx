/* eslint-disable promise/always-return */
/* eslint-disable @typescript-eslint/ban-ts-comment */
/* eslint-disable react/jsx-props-no-spreading */
/* eslint-disable import/prefer-default-export */
/* eslint-disable no-use-before-define */
import clsx from 'clsx';
import { AnimatePresence } from 'framer-motion';
import React, { useCallback, useEffect, useState } from 'react';

import { Browser } from 'renderer/App/components/Browser';
import { BrowserProps } from 'renderer/App/components/Browser/Types';
import { useBoard } from 'renderer/App/hooks/useBoard';
import { useBrowserMethods } from 'renderer/App/hooks/useBrowserMethods';
import { useStoreHelpers } from 'renderer/App/hooks/useStoreHelpers';
import { useAppDispatch } from 'renderer/App/store/hooks';
import {
  setActiveBrowser,
  setBoardHeight,
} from 'renderer/App/store/reducers/Board';

import { BoardProps } from './Types';

import './style.scss';

export const Board: React.FC<BoardProps> = ({ isFullSize, boardId }) => {
  const board = useBoard();
  const dispatch = useAppDispatch();
  const { focus } = useBrowserMethods();
  const [items, setItems] = useState<BrowserProps[]>([]);
  const helpers = useStoreHelpers();

  const boardContainer = document.querySelector('#Board__container');

  const makeBrowsers = useCallback((sorted: BrowserProps[]) => {
    return sorted.map((b) => <Browser {...b} key={b.id} firstRendering />);
  }, []);

  const contextMenuListener = (e: MouseEvent) => {
    e.preventDefault();
    window.app.tools.showBoardContextMenu({ x: e.clientX, y: e.clientY });
  };

  useEffect(() => {
    if (!board.isFullSize) {
      setTimeout(() => {
        helpers.board
          .distributeWindowsByOrder(board.browsers)
          .then(() => {
            setTimeout(
              () => board.activeBrowser && focus(board.activeBrowser),
              300
            );
          })
          .catch(console.log);
      }, 300);
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [board.isFullSize]);

  useEffect(() => {
    const toSort = [...board.browsers];
    const sorted = toSort.sort((a, b) => {
      return a.id > b.id ? 1 : -1;
    });
    setItems(sorted);

    if (boardContainer) {
      const max = board.browsers.reduce(
        (acc, val) => Math.max(acc, val.top + val.height + 100),
        0
      );

      // @ts-ignore
      boardContainer.style.height = `${Number(max)}px`;

      dispatch(setBoardHeight(max));
    }
  }, [board.browsers, dispatch, boardContainer]);

  useEffect(() => {
    window.app.board.setWindowsCount({
      boardId: board.id,
      count: board.browsers.length,
    });
  }, [board.browsers.length, board.id]);

  useEffect(() => {
    if (board.activeBrowser) {
      focus(board.activeBrowser, true);
      dispatch(setActiveBrowser(board.activeBrowser));
    }
  }, [board.activeBrowser, focus, dispatch]);

  // focus activeBrowser when componentDidMount
  useEffect(() => {
    setTimeout(() => {
      if (board.activeBrowser) focus(board.activeBrowser);
      if (boardId) {
        window.app.browser
          .getUrlToOpen()
          .then((res) => {
            if (res) helpers.browser.add({ url: res });
          })
          .catch(console.log);
      }
    }, 0);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  useEffect(() => {
    document
      .getElementById('Board__container')
      ?.addEventListener('contextmenu', contextMenuListener);

    return () =>
      document
        .getElementById('Board__container')
        ?.removeEventListener('contextmenu', contextMenuListener);
  }, []);

  return (
    <div
      id="Board__container"
      className={clsx({
        'Board__is-maximized': board.isFullSize || isFullSize,
        'Board__isnt-maximized': !board.isFullSize && !isFullSize,
      })}
    >
      <AnimatePresence>{makeBrowsers(items)}</AnimatePresence>
      <div className="Board__edge-snap-left" />
      <div className="Board__edge-snap-right" />
      <div className="Board__edge-snap-maximized" />
    </div>
  );
};
