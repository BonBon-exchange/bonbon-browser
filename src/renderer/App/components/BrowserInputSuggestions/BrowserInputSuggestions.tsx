/* eslint-disable @typescript-eslint/no-unused-expressions */
/* eslint-disable import/prefer-default-export */
/* eslint-disable promise/always-return */
/* eslint-disable jsx-a11y/click-events-have-key-events */
/* eslint-disable jsx-a11y/no-noninteractive-element-interactions */
import { useEffect, useState, useCallback } from 'react';

import { BrowserInputSuggestionsProps, SuggestionItem } from './Types';

import './style.scss';

export const BrowserInputSuggestions: React.FC<
  BrowserInputSuggestionsProps
> = ({
  inputValue,
  handleSuggestionClick,
  setSelectedSuggestion,
  setDomainSuggestionResults,
}: BrowserInputSuggestionsProps) => {
  const [suggestions, setSuggestions] = useState<SuggestionItem[]>([]);
  const [selected, setSelected] = useState<SuggestionItem | null>(null);

  const keyDownListener = useCallback(
    (e: KeyboardEvent) => {
      let direction: 'next' | 'previous' | null;
      const selectedCurrentIndex = suggestions.findIndex((s) => s === selected);

      switch (e.key) {
        case 'ArrowDown':
          direction = 'next';
          break;

        case 'ArrowUp':
          direction = 'previous';
          break;

        default:
          direction = null;
          if (selected && selectedCurrentIndex === -1) setSelected(null);
          break;
      }

      if (direction) {
        if (selectedCurrentIndex === -1) {
          direction === 'next'
            ? setSelected(suggestions[0])
            : setSelected(suggestions[suggestions.length - 1]);
        } else {
          let newIndex;
          direction === 'next'
            ? (newIndex = selectedCurrentIndex + 1)
            : (newIndex = selectedCurrentIndex - 1);

          if (newIndex > suggestions.length - 1) {
            newIndex = 0;
          }

          if (newIndex < 0) {
            newIndex = suggestions.length - 1;
          }

          setSelected(suggestions[newIndex]);
        }
      }
    },
    [selected, suggestions]
  );

  useEffect(() => {
    if (selected) {
      switch (selected.type) {
        case 'domain':
          setSelectedSuggestion(selected?.display || null);
          break;

        default:
          setSelectedSuggestion(selected?.url || null);
          break;
      }
    }
  }, [setSelectedSuggestion, selected]);

  useEffect(() => {
    if (inputValue.length === 0) {
      setSuggestions([]);
      setSelected(null);
      return;
    }

    const promises = [
      window.app.bookmark.findInBookmarks(inputValue),
      window.app.history.findInHistory(inputValue),
      window.app.tools.findInKnownDomains(inputValue),
    ];

    Promise.all(promises)
      .then((result: any[]) => {
        // 2 suggestions of bookmarks
        const rows = result[0].slice(0, 2).map((r: SuggestionItem) => ({
          ...r,
          type: 'bookmark',
          id: `bookmark::${r.id}`,
        }));

        // 4 suggestions of domains
        const domainsResults: SuggestionItem[] = result[2]
          .map((r: any, i: number) => {
            return {
              id: `domain::${r.id}::${i}`,
              display: r.domain,
              type: 'domain',
            };
          })
          .reduce((acc: SuggestionItem[], val: SuggestionItem) => {
            const exist = acc.find((a) => a.display === val.display);
            return exist ? acc : [...acc, val];
          }, [])
          .slice(0, 3);

        setDomainSuggestionResults(domainsResults);
        rows.push(...domainsResults);

        // 2 suggestions of history
        rows.push(
          ...result[1].slice(0, 2).map((r: SuggestionItem) => ({
            ...r,
            type: 'history',
            id: `history::${r.id}`,
          }))
        );

        setSuggestions(rows || []);
      })
      .catch(console.log);
  }, [inputValue, setDomainSuggestionResults]);

  useEffect(() => {
    window.addEventListener('keydown', keyDownListener);
    return () => window.removeEventListener('keydown', keyDownListener);
  }, [keyDownListener]);

  return (
    <div className="BrowserInputSuggestions__container">
      <ul>
        {suggestions.map((s) => (
          <li
            key={`${s.id}::${s.url}`}
            className={
              s === selected
                ? 'BrowserInputSuggestions__selected-item'
                : undefined
            }
            onClick={() => handleSuggestionClick(s.url || s.display || '')}
          >
            {s.display || s.url}
          </li>
        ))}
      </ul>
    </div>
  );
};
