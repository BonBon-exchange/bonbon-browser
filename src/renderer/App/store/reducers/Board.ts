/* eslint-disable @typescript-eslint/no-unused-expressions */
import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { v4 } from 'uuid';

import { WritableDraft } from 'immer/dist/internal';
import { BoardType } from 'renderer/App/components/Board/Types';
import { BrowserProps } from 'renderer/App/components/Browser/Types';

type UpdateBrowserType = {
  browserId: string;
  params: Partial<BrowserProps>;
};

type UpdateBrowserUrlType = {
  url: string;
  browserId: string;
};

type UpdateBrowserLoadingType = {
  isLoading: boolean;
  browserId: string;
};

type UpdateBrowserTitleType = {
  title: string;
  browserId: string;
};

type UpdateBrowserFavType = {
  favicon: string;
  browserId: string;
};

type UpdateBrowserCertificateFingerprintType = {
  certificateErrorFingerprint: string | null;
  browserId: string;
};

interface BoardState {
  board: BoardType;
}

const boardId = v4();
const newBoard = {
  id: boardId,
  label: 'New collection',
  browsers: [],
  activeBrowser: null,
  closedUrls: [],
  isFullSize: false,
  browsersActivity: [],
  height: 0,
};

export const initialState: BoardState = {
  board: newBoard,
};

const addCLosedUrl = (state: WritableDraft<BoardState>, url: string) => {
  if (state.board.closedUrls.length > 20) {
    state.board.closedUrls.splice(0, 1);
  }
  state.board.closedUrls.push(url);
};

const addBrowserActivity = (
  state: WritableDraft<BoardState>,
  browserId: string
) => {
  if (state.board.browsersActivity.length > 20) {
    state.board.browsersActivity.splice(0, 1);
  }
  state.board.browsersActivity = state.board.browsersActivity.filter(
    (b) => b !== browserId
  );
  state.board.browsersActivity.push(browserId);
};

const findAndSelectTabForExtensionsFromBrowserId = (
  state: WritableDraft<BoardState>,
  browserId: string
): void => {
  const wcId = state.board.browsers.find(
    (b) => b.id === browserId
  )?.webContentsId;
  if (wcId) window.app.browser.select(wcId);
};

const setStateActiveBrowser = (
  state: WritableDraft<BoardState>,
  browserId: string
): void => {
  state.board.activeBrowser = browserId;
  addBrowserActivity(state, browserId);
  findAndSelectTabForExtensionsFromBrowserId(state, browserId);
};

const getBrowserIndexFromBrowserId = (
  state: WritableDraft<BoardState>,
  browserId: string
): number => {
  return state.board.browsers.findIndex((b) => b.id === browserId);
};

export const boardSlice = createSlice({
  name: 'board',
  initialState,
  reducers: {
    setBoard: (state, action: PayloadAction<BoardType>) => {
      state.board = action.payload;
      const activeBrowser = state.board.browsers.find(
        (b) => b.id === state.board.activeBrowser
      );
      if (!state.board.activeBrowser || !activeBrowser) {
        if (state.board.browsers.length > 0) {
          setStateActiveBrowser(
            state,
            state.board.browsers[state.board.browsers.length - 1].id
          );
        }
      }
      if (state.board.activeBrowser) {
        findAndSelectTabForExtensionsFromBrowserId(
          state,
          state.board.activeBrowser
        );
      }
    },
    toggleBoardFullSize: (state) => {
      state.board.isFullSize = !state.board.isFullSize;
      window.app.analytics.event(`toggle_fullsize_${!state.board.isFullSize}`);
    },
    setActiveBrowser: (state, action: PayloadAction<string>) => {
      setStateActiveBrowser(state, action.payload);
    },
    addBrowser: (state, action: PayloadAction<BrowserProps>) => {
      if (action.payload.browserIndex) {
        state.board.browsers.splice(
          action.payload.browserIndex,
          0,
          action.payload
        );
      } else {
        state.board.browsers.push(action.payload);
      }
      setStateActiveBrowser(state, action.payload.id);
      window.app.analytics.event('add_browser');
    },
    setBrowsers: (state, action: PayloadAction<BrowserProps[]>) => {
      state.board.browsers = action.payload;
    },
    updateBrowser: (state, action: PayloadAction<UpdateBrowserType>) => {
      const browserIndex = getBrowserIndexFromBrowserId(
        state,
        action.payload.browserId
      );
      if (browserIndex > -1) {
        state.board.browsers[browserIndex] = {
          ...state.board.browsers[browserIndex],
          ...action.payload.params,
        };
      }
    },
    updateBrowserUrl: (state, action: PayloadAction<UpdateBrowserUrlType>) => {
      const browserIndex = getBrowserIndexFromBrowserId(
        state,
        action.payload.browserId
      );
      if (browserIndex > -1) {
        state.board.browsers[browserIndex].url = action.payload.url;
      }
    },
    updateBrowserLoading: (
      state,
      action: PayloadAction<UpdateBrowserLoadingType>
    ) => {
      const browserIndex = getBrowserIndexFromBrowserId(
        state,
        action.payload.browserId
      );
      if (browserIndex > -1) {
        state.board.browsers[browserIndex].isLoading = action.payload.isLoading;
      }
    },
    updateBrowserTitle: (
      state,
      action: PayloadAction<UpdateBrowserTitleType>
    ) => {
      const browserIndex = getBrowserIndexFromBrowserId(
        state,
        action.payload.browserId
      );
      if (browserIndex > -1) {
        state.board.browsers[browserIndex].title = action.payload.title;
      }
    },
    updateBrowserFav: (state, action: PayloadAction<UpdateBrowserFavType>) => {
      const browserIndex = getBrowserIndexFromBrowserId(
        state,
        action.payload.browserId
      );
      if (browserIndex > -1) {
        state.board.browsers[browserIndex].favicon = action.payload.favicon;
      }
    },
    renameBoard: (state, action: PayloadAction<string>) => {
      state.board.label = action.payload;
      window.app.analytics.event('rename_board');
    },
    minimizeBrowser: (state, action: PayloadAction<string>) => {
      const browserIndex = getBrowserIndexFromBrowserId(state, action.payload);
      state.board.browsers[browserIndex].isMinimized = true;
    },
    unminimizeBrowser: (state, action: PayloadAction<string>) => {
      const browserIndex = getBrowserIndexFromBrowserId(state, action.payload);

      state.board.browsers[browserIndex].isMinimized = false;
    },
    removeBrowser: (state, action: PayloadAction<string>) => {
      const browserIndex = getBrowserIndexFromBrowserId(state, action.payload);

      // remove browser from state
      if (browserIndex > -1) {
        if (!state.board.closedUrls) state.board.closedUrls = [];
        addCLosedUrl(state, state.board.browsers[browserIndex].url);
        state.board.lastClosedBrowserDimensions = [
          state.board.browsers[browserIndex].width,
          state.board.browsers[browserIndex].height,
        ];
        state.board.browsers.splice(browserIndex, 1);

        // clean browsersActivity state
        state.board.browsersActivity = state.board.browsersActivity.filter(
          (b) => b !== action.payload
        );
      }

      // clean activeBrowser
      if (state.board.activeBrowser === action.payload) {
        if (state.board.browsers.length > 0) {
          setStateActiveBrowser(
            state,
            state.board.browsersActivity[
              state.board.browsersActivity.length - 1
            ]
          );
        } else {
          state.board.activeBrowser = null;
          window.app.browser.selectBrowserView();
        }
      }

      window.app.analytics.event('close_browser');
    },
    removeAllBrowsersExcept: (state, action: PayloadAction<string>) => {
      state.board.browsers
        .filter((b) => b.id !== action.payload)
        .forEach((b) => addCLosedUrl(state, b.url));

      state.board.browsers = state.board.browsers.filter(
        (b) => b.id === action.payload
      );

      // clean activeBrowser
      setStateActiveBrowser(state, action.payload);

      // clean browsersActivity state
      state.board.browsersActivity = state.board.browsersActivity.filter(
        (b) => b === action.payload
      );

      // send event
      window.app.analytics.event('close_others_browser');
    },
    updateBrowserCertificateErrorFingerprint: (
      state,
      action: PayloadAction<UpdateBrowserCertificateFingerprintType>
    ) => {
      const browserIndex = getBrowserIndexFromBrowserId(
        state,
        action.payload.browserId
      );
      if (browserIndex > -1) {
        state.board.browsers[browserIndex].certificateErrorFingerprint =
          action.payload.certificateErrorFingerprint;
      }
    },
    removeAllBrowsers: (state) => {
      if (!state.board.closedUrls) state.board.closedUrls = [];
      state.board.browsers.forEach((b) => addCLosedUrl(state, b.url));
      state.board.browsers = [];
      state.board.browsersActivity = [];
      state.board.activeBrowser = null;
      window.app.browser.selectBrowserView();
      window.app.analytics.event('close_all_browsers');
    },
    removeLastClosedUrl: (state) => {
      state.board.closedUrls.splice(state.board.closedUrls.length - 1, 1);
    },
    toggleSearch: (state, action: PayloadAction<string>) => {
      const browserIndex = getBrowserIndexFromBrowserId(state, action.payload);
      state.board.browsers[browserIndex].isSearching =
        !state.board.browsers[browserIndex].isSearching;
    },
    setLastResizedBrowserDimensions: (
      state,
      action: PayloadAction<[number, number]>
    ) => {
      state.board.lastResizedBrowserDimensions = action.payload;
    },
    setBoardHeight: (state, action: PayloadAction<number>) => {
      state.board.height = action.payload;
    },
  },
});

export const {
  setBoard,
  updateBrowserUrl,
  updateBrowserFav,
  renameBoard,
  removeBrowser,
  minimizeBrowser,
  unminimizeBrowser,
  removeAllBrowsers,
  addBrowser,
  setActiveBrowser,
  updateBrowserTitle,
  updateBrowser,
  toggleBoardFullSize,
  removeLastClosedUrl,
  setBrowsers,
  removeAllBrowsersExcept,
  updateBrowserLoading,
  updateBrowserCertificateErrorFingerprint,
  toggleSearch,
  setLastResizedBrowserDimensions,
  setBoardHeight,
} = boardSlice.actions;

export default boardSlice.reducer;
