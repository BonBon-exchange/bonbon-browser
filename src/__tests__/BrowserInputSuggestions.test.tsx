/* eslint-disable react/jsx-props-no-spreading */
import '@testing-library/jest-dom';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import { Middleware } from '@reduxjs/toolkit';
import renderer, { act } from 'react-test-renderer';

import { BrowserInputSuggestions } from '../renderer/App/components/BrowserInputSuggestions';
import { mockWindow } from './beforeAll';
import { initialState } from '../renderer/App/store/reducers/Board';

let tree: any;
let store: any;
const middlewares: Middleware[] = [];

const props = {
  inputValue: 'dani',
  handleSuggestionClick: jest.fn(),
  setSelectedSuggestion: jest.fn(),
  setDomainSuggestionResults: jest.fn(),
};

describe('BrowserInputSuggestions', () => {
  beforeAll(() => {
    mockWindow();
    const mockStore = configureStore(middlewares);
    store = mockStore({ board: initialState });
  });

  it('should render', () => {
    act(() => {
      tree = renderer.create(
        <Provider store={store}>
          <BrowserInputSuggestions {...props} />
        </Provider>
      );
    });

    expect(tree).toMatchSnapshot();
  });
});
