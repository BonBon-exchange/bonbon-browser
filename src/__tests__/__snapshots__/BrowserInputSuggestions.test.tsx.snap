// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BrowserInputSuggestions should render 1`] = `
<div
  className="BrowserInputSuggestions__container"
>
  <ul />
</div>
`;
